### ISSUE TITLE
Django-role-permission duplicate queries

### ISSUE THREAD

**Issue by kheshav** [vintasoftware/django-role-permissions#81]
Hello,

Am experiencing duplicate queries for function `has_permission ` 

I have a decorator as follows:

```
def canAccessDashboard(func):
    def func_wrapper(*args):
        user = cache.get('UserObj_' + str(args[0].user.id))
        if user is None:
            user = User.objects.get(id=args[0].user.id)
            cache.set('UserObj_' + str(args[0].user.id), user, 300)
        if has_permission(user, "access_dasboard"):
            if not has_permission(user, "add_merchant_address"):
                return render(args[0], 'dashboard/specify_address.html')
            return func(*args)
        return redirect(settings.VERIFY_REGISTRATION_URL)
    return func_wrapper
```

Which is called on my django view as decorator:

```
@canAccessDashboard
def index(request):
    """ Merchant settings main page"""
    try:
        return render(request, 'dashboard/merchantsettings.html')
    except Exception as e:
        return HttpResponse('NOK', status=500)
```

The issue is that am experiencing  duplicate queries for each permission in my available_permission:

```
available_permissions = {
        'access_dasboard': False,
        'add_merchant_address': False,
        'immediate_cashout': False,
        'update_merchant_address': False
    }
```

> SELECT ••• FROM "auth_permission" INNER JOIN "auth_user_user_permissions" ON ("auth_permission"."id" = "auth_user_user_permissions"."permission_id") INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE "auth_user_user_permissions"."user_id" = 16 ORDER BY "django_content_type"."app_label" ASC, "django_content_type"."model" ASC, "auth_permission"."codename" ASC
> Duplicated 8 times.
> 
> SELECT ••• FROM "auth_permission" WHERE ("auth_permission"."content_type_id" = 4 AND "auth_permission"."codename" = 'add_merchant_address')
> Duplicated 8 times
> 
> SELECT ••• FROM "auth_permission" WHERE ("auth_permission"."content_type_id" = 4 AND "auth_permission"."codename" = 'immediate_cashout')
> Duplicated 8 times.
> 
> SELECT ••• FROM "auth_permission" WHERE ("auth_permission"."content_type_id" = 4 AND "auth_permission"."codename" = 'access_dasboard') Duplicated 8 times
> 

Can you please fix it out?

Thank you in advance.

**Comment by lucabezerra** [comment]
This is completely unrelated to the issue itself, but just a heads up, it seems that you've got a typo in your permission's name at `if has_permission(user, "access_dasboard"):` :)

Someone should be looking at this issue and providing you a proper answer soon.

**Comment by filipeximenes** [comment]
@kheshav thanks for reporting! Feel free to work on this performance improvement, I'm happy to assist you with the PR process!
The fix will be probably in this function: https://github.com/vintasoftware/django-role-permissions/blob/master/rolepermissions/permissions.py#L52

**Comment by kheshav** [comment]
Hello @filipeximenes , thanks for your feedback. I have made a pull request regarding the Issue, [PR 82](https://github.com/vintasoftware/django-role-permissions/pull/82) 

The Fix solved the issues of duplicate queries on my side.  :)

Please check and Merge if so required.

Thanks.

**Comment by aldookware** [comment]
I there a solution to this problem?

**Comment by powderflask** [comment]
This can be a serious performance hit, which can't be resolved be simply pre-fetching related groups and permissions.   
`available_perm_status(user)` makes a query for EACH permission the user has.  I have one 'dashboard' type view that checks about 20 object permissions, each of which might check 2-3 user permissions (e.g., access_any, access_own, etc.) and Users that typically have 10-20 permissions.  When you do the math, that creates hundreds of redundant queries.
`get_user_roles(user)` has a similar performance issue, but not nearly as severe and much easier to resolve.
Ideally, the solution to this issue would result in ZERO queries when the user object has groups and permissions pre-fetched, and TWO queries (one for roles and one for permissions) otherwise.  I believe this means the solution must fetch `all()` related groups / permissions, and do filtering in Python (https://docs.djangoproject.com/en/dev/ref/models/querysets/#django.db.models.query.QuerySet.prefetch_related)

I have hacked in a solution that has the above properties and meets my needs.  Will work on creating a simple test case that demonstrates the issue, and a PR with a more polished version of this solution.
Since this code does not introduce any observable API behaviour, I assume the only new test required will verify that no extra queries are generated?

**Comment by powderflask** [comment]
Please see PR #93 for my ideas for resolving this.

**Comment by filipeximenes** [comment]
Thanks for the great work @powderflask!

**Comment by filipeximenes** [comment]
Released on PyPI as version 2.2.1
